x-pack/plugin/core/src/main/java/org/elasticsearch/index/engine/FrozenEngine.java

```diff
@@ -0,0 +1,520 @@
+        DirectoryReader reader = null;
+        boolean success = false;
+        try {
+            reader = getReader();
+            if (reader == null) {
+                for (ReferenceManager.RefreshListener listeners : config ().getInternalRefreshListener()) {
+                    listeners.beforeRefresh();
+                }
+                reader = DirectoryReader.open(engineConfig.getStore().directory());
+                searcherFactory.processReaders(reader, null);
+                reader = lastOpenedReader = wrapReader(reader, Function.identity());
+                reader.getReaderCacheHelper().addClosedListener(this::onReaderClosed);
+                for (ReferenceManager.RefreshListener listeners : config ().getInternalRefreshListener()) {
+                    listeners.afterRefresh(true);
+                }
+            }
+            success = true;
+            return reader;
+        } finally {
+            if (success == false) {
+                IOUtils.close(reader);
+            }
+        }
+    }
+
+    @SuppressForbidden(reason = "we manage references explicitly here")
+    private synchronized DirectoryReader getReader() throws IOException {
+        DirectoryReader reader = null;
+        boolean success = false;
+        try {
+            if (lastOpenedReader != null && lastOpenedReader.tryIncRef()) {
+                reader = lastOpenedReader;
+            }
+            success = true;
+            return reader;
+        } finally {
+            if (success == false) {
+                IOUtils.close(reader);
+            }
+        }
+    }
+
+    @Override
+    @SuppressWarnings("fallthrough")
+    @SuppressForbidden( reason = "we manage references explicitly here")
+    public Searcher acquireSearcher(String source, SearcherScope scope) throws EngineException {
+        store.incRef();
+        boolean success = false;
+        try  {
+            final boolean maybeOpenReader;
+            switch (source) {
+                case "load_seq_no":
+                case "load_version":
+                    assert false : "this is a read-only engine";
+                case "doc_stats":
+                    assert false : "doc_stats are overwritten";
+                case "segments":
+                case "segments_stats":
+                case "completion_stats":
+                case "refresh_needed":
+                    maybeOpenReader = false;
+                    break;
+                default:
+                    maybeOpenReader = true;
+            }
+            // special case we only want to report segment stats if we have a reader open. in that case we only get a reader if we still
+            // have one open at the time and can inc it's reference.
+            DirectoryReader reader = maybeOpenReader ? getOrOpenReader() : getReader();
+            if (reader == null) {
+                store.decRef();
+                success = true;
```

Comment:
why do we need to decrement the store ref early?

Reply:
we don't I can fix that to make it more intuitive